data_structures/hashing/bloom_filter.py

```diff
@@ -0,0 +1,105 @@
+"""
+See https://en.wikipedia.org/wiki/Bloom_filter
+"""
+from hashlib import md5, sha256
+from random import choices
+from string import ascii_lowercase
+
+
+class Bloom:
+    # number of hash functions is fixed
+    HASH_FUNCTIONS = (sha256, md5)
+
+    def __init__(self, size: int = 8) -> None:
+        self.bitstring = 0b0
+        self.size = size
+
+    def add(self, value: str) -> None:
+        h = self.hash_(value)
+        self.bitstring |= h
+        print(
+            f"""\
+[add] value =      {value}
+      hash =       {self.format_bin(h)}
+      filter =     {self.format_bin(self.bitstring)}
+"""
+        )
+
+    def exists(self, value: str) -> bool:
+        h = self.hash_(value)
+        res = (h & self.bitstring) == h
+
+        print(
+            f"""\
+[exists] value =   {value}
+         hash =    {self.format_bin(h)}
+         filter =  {self.format_bin(self.bitstring)}
+         res =     {res}
+"""
+        )
+        return res
+
+    def format_bin(self, value: int) -> str:
+        res = bin(value)[2:]
+        return res.zfill(self.size)
+
+    def hash_(self, value: str) -> int:
```

Comment:
```suggestion
    def hash(self, value: str) -> int:
```
Not sure if the preceding underscore is necessary here - Yes `hash` is a built in function, however this function is binded to the class

Reply:
I though the same, the `hash` hash built-in function is still accessible.

I put the underscore because but ruff complains:
![2023-04-07-173709_](https://user-images.githubusercontent.com/10489506/230636795-ed99af47-876c-48a0-8877-e4c9ef70ed46.png)

Reply:
@CaedenPH The current code follows ruff's advice which I believe is less confusing to readers..